File: patterns/09_repository/csharp/Tabularium.cs
// ============================================================
//  REPOSITORY PATTERN — The Imperial Archive
//  Run: dotnet script Tabularium.cs
// ============================================================
#nullable enable
using System;
using System.Collections.Generic;
using System.Linq;

// ── Domain Object ────────────────────────────────────────────
record Centurion(string Id, string Name, string Legion, string Rank, int BattlesWon);

// ── Repository Interface ─────────────────────────────────────
interface ICenturionRepo {
    Centurion? FindById(string id);
    List<Centurion> FindByLegion(string legion);
    Centurion Save(Centurion c);
    bool Delete(string id);
}

// ── In-Memory Implementation ─────────────────────────────────
class InMemoryRepo : ICenturionRepo {
    private readonly Dictionary<string, Centurion> _store = new Dictionary<string, Centurion>();

    public Centurion? FindById(string id) {
        Centurion? c;
        return _store.TryGetValue(id, out c) ? c : null;
    }

    public List<Centurion> FindByLegion(string legion) =>
        _store.Values.Where(c => c.Legion == legion).ToList();

    public Centurion Save(Centurion c) {
        _store[c.Id] = c;
        Console.WriteLine($"  📚 Saved: {c.Name} ({c.Rank}) [{c.Id}]");
        return c;
    }

    public bool Delete(string id) => _store.Remove(id);
}

// ── Service (business logic) ──────────────────────────────────
class CenturionService {
    private readonly ICenturionRepo _repo;
    public CenturionService(ICenturionRepo repo) { _repo = repo; }

    public string PromoteBest(string legion) {
        var all = _repo.FindByLegion(legion);
        if (all.Count == 0) return $"No centurions in {legion}";
        var best = all.OrderByDescending(c => c.BattlesWon).First();
        _repo.Save(best with { Rank = "Primus Pilus" });
        return $"🎖  {best.Name} promoted! ({best.BattlesWon} battles won)";
    }
}

// ── Demo ──────────────────────────────────────────────────────
Console.WriteLine("╔═══════════════════════════════════════════════╗");
Console.WriteLine("║   REPOSITORY PATTERN — C#                     ║");
Console.WriteLine("║   Tabularium Capitolinum                      ║");
Console.WriteLine("╚═══════════════════════════════════════════════╝\n");

var repo = new InMemoryRepo();
repo.Save(new Centurion("c1", "Marcus Aurelius",  "Legio X", "Centurion", 15));
repo.Save(new Centurion("c2", "Gaius Petronius",  "Legio X", "Centurion", 23));
repo.Save(new Centurion("c3", "Titus Labienus",   "Legio X", "Centurion", 31));
repo.Save(new Centurion("c4", "Julia Domna",      "Legio V", "Centurion", 19));

Console.WriteLine("\n── QUERY: All of Legio X ───────────────────────────");
foreach (var c in repo.FindByLegion("Legio X"))
    Console.WriteLine($"  {c.Name,-20} {c.BattlesWon,2} battles  [{c.Rank}]");

Console.WriteLine("\n── PROMOTE BEST ────────────────────────────────────");
var svc = new CenturionService(repo);
Console.WriteLine("  " + svc.PromoteBest("Legio X"));

Console.WriteLine("\n── FIND BY ID ──────────────────────────────────────");
var found = repo.FindById("c1");
if (found != null)
    Console.WriteLine($"  Found: {found.Name} — now [{found.Rank}]");

Console.WriteLine("\n\"Ubi data est? Tabularius scit!\"");

File: patterns/09_repository/cpp/tabularium.cpp
// ============================================================
//  REPOSITORY PATTERN — The Imperial Archive (Tabularium)
//  "Ubi data est? Tabularius scit"
//  Where is the data? The archivist knows.
//
//  Compile:  g++ -std=c++17 -o tabularium tabularium.cpp
//  Run:      ./tabularium
// ============================================================
#include <iostream>
#include <map>
#include <vector>
#include <string>
#include <optional>
#include <algorithm>
using namespace std;

struct Centurion {
    string id, name, legion, rank;
    int battlesWon;
};

class ICenturionRepository {
public:
    virtual ~ICenturionRepository() = default;
    virtual optional<Centurion>  findById(const string& id)        = 0;
    virtual vector<Centurion>    findByLegion(const string& legion) = 0;
    virtual vector<Centurion>    findAll()                          = 0;
    virtual void                 save(const Centurion& c)           = 0;
    virtual bool                 remove(const string& id)           = 0;
};

class InMemoryCenturionRepo : public ICenturionRepository {
    map<string,Centurion> _store;
public:
    optional<Centurion> findById(const string& id) override {
        auto it = _store.find(id);
        return it != _store.end() ? optional<Centurion>{it->second} : nullopt;
    }
    vector<Centurion> findByLegion(const string& l) override {
        vector<Centurion> res;
        for (auto& [k,v] : _store) if (v.legion==l) res.push_back(v);
        return res;
    }
    vector<Centurion> findAll() override {
        vector<Centurion> res;
        for (auto& [k,v] : _store) res.push_back(v);
        return res;
    }
    void save(const Centurion& c) override {
        _store[c.id] = c;
        cout << "  📚 Saved: " << c.name << " (" << c.rank << ") [" << c.id << "]\n";
    }
    bool remove(const string& id) override { return _store.erase(id) > 0; }
};

class CenturionService {
    ICenturionRepository& _repo;
public:
    CenturionService(ICenturionRepository& r) : _repo(r) {}
    void promoteBest(const string& legion) {
        auto all = _repo.findByLegion(legion);
        if (all.empty()) { cout << "  No centurions in " << legion << "\n"; return; }
        auto best = *max_element(all.begin(), all.end(),
            [](const Centurion& a, const Centurion& b){ return a.battlesWon < b.battlesWon; });
        best.rank = "Primus Pilus";
        _repo.save(best);
        cout << "  🎖  " << best.name << " promoted to Primus Pilus! (" << best.battlesWon << " battles)\n";
    }
};

int main() {
    cout << "╔═══════════════════════════════════════════════╗\n";
    cout << "║   REPOSITORY PATTERN — C++                    ║\n";
    cout << "║   The Imperial Archive (Tabularium)           ║\n";
    cout << "╚═══════════════════════════════════════════════╝\n\n";

    InMemoryCenturionRepo repo;
    repo.save({"c1","Marcus Aurelius",  "Legio X","Centurion",15});
    repo.save({"c2","Gaius Petronius",  "Legio X","Centurion",23});
    repo.save({"c3","Publius Quinctius","Legio XII","Centurion",7});
    repo.save({"c4","Titus Labienus",   "Legio X","Centurion",31});

    cout << "\n── QUERY: All Legio X Centurions ───────────────\n";
    for (auto& c : repo.findByLegion("Legio X"))
        cout << "  " << c.name << " — " << c.battlesWon << " battles\n";

    cout << "\n── SERVICE: Promote Best in Legio X ────────────\n";
    CenturionService svc(repo);
    svc.promoteBest("Legio X");

    cout << "\n── VERIFY PROMOTION ────────────────────────────\n";
    auto found = repo.findById("c4");
    if (found) cout << "  " << found->name << " is now: " << found->rank << "\n";

    cout << "\n\"Ubi data est? Tabularius scit!\"\n";
    return 0;
}
